src/tools/rbd/action/List.cc

```diff
@@ -4,25 +4,182 @@
+      f->dump_string("pool", pool);
+      f->dump_string("image", image);
+      f->dump_string("snapshot", snap);
+      f->close_section();
+    }
+    f->dump_int("format", old_format ? 1 : 2);
+    if (!lockers.empty())
+      f->dump_string("lock_type", exclusive ? "exclusive" : "shared");
+    f->close_section();
+  } else {
+    tbl << w->name
+        << stringify(si_t(info.size))
+        << parent
+        << ((old_format) ? '1' : '2')
+        << ""                         // protect doesn't apply to images
+        << lockstr
+        << TextTable::endrow;
+  }
+
+  std::vector<librbd::snap_info_t> snaplist;
+  if (w->img.snap_list(snaplist) >= 0 && !snaplist.empty()) {
+    for (std::vector<librbd::snap_info_t>::iterator s = snaplist.begin();
+         s != snaplist.end(); ++s) {
+      bool is_protected;
+      bool has_parent = false;
+      parent.clear();
+      w->img.snap_set(s->name.c_str());
+      r = w->img.snap_is_protected(s->name.c_str(), &is_protected);
+      if (r < 0)
+        return r;
+      if (w->img.parent_info(&pool, &image, &snap) >= 0) {
+        parent = pool + "/" + image + "@" + snap;
+        has_parent = true;
+      }
+      if (f) {
+        f->open_object_section("snapshot");
+        f->dump_string("image", w->name);
+        f->dump_string("snapshot", s->name);
+        f->dump_unsigned("size", s->size);
+        if (has_parent) {
+          f->open_object_section("parent");
+          f->dump_string("pool", pool);
+          f->dump_string("image", image);
+          f->dump_string("snapshot", snap);
+          f->close_section();
+        }
+        f->dump_int("format", old_format ? 1 : 2);
+        f->dump_string("protected", is_protected ? "true" : "false");
+        f->close_section();
+      } else {
+        tbl << w->name + "@" + s->name
+            << stringify(si_t(s->size))
+            << parent
+            << ((old_format) ? '1' : '2')
+            << (is_protected ? "yes" : "")
+            << ""                     // locks don't apply to snaps
+            << TextTable::endrow;
+      }
+    }
+  }
+
+  return r < 0 ? r : 0;
+}
+
+void init_worker(worker_entry* worker, librbd::RBD* rbd, librados::IoCtx* ioctx)
+{
+  worker->io_ctx = ioctx;
+  worker->rbd = rbd;
+  worker->opened = 0;
+}
+
+int do_list(std::string &pool_name, bool lflag, int threads, Formatter *f) {
+  std::vector<worker_entry*> workers;
   std::vector<std::string> names;
-  int r = rbd.list(io_ctx, names);
+  librados::Rados rados;
+  librbd::RBD rbd;
+  librados::IoCtx ioctx;
+
+  if (threads < 1) {
+    threads = 1;
+  }
+  if (threads > 32) {
+    threads = 32;
+  }
+
+  int r = rados.init_with_context(g_ceph_context);
+  if (r < 0) {
+    std::cerr << "rbd: couldn't initialize rados!" << std::endl;
+    return r;
+  }
+
+  r = rados.connect();
+  if (r < 0) {
+    std::cerr << "rbd: couldn't connect to the cluster!" << std::endl;
+    return r;
+  }
+
+  worker_entry* main = new worker_entry();
```

Comment:
Nit: this worker entry doesn't serve much value -- just directly use the ```librbd::RBD``` object, etc when retrieving the image list.